tcpsocket.8
.\" Automatically generated by Pod::Man version 1.04
.\" Thu Jan 18 18:45:31 2001
.\"
.\" Standard preamble:
.\" ======================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R

.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used
.\" to do unbreakable dashes and therefore won't be available.  \*(C` and
.\" \*(C' expand to `' in nroff, nothing in troff, for use with C<>
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` `
.    ds C' '
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr
.\" for titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and
.\" index entries marked with X<> in POD.  Of course, you'll have to process
.\" the output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
.    .
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it
.\" makes way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.bd B 3
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ======================================================================
.\"
.IX Title "TCPSOCKET 1"
.TH TCPSOCKET 1 "0.2.4" "2001-01-18" "iodp"
.UC
.SH "NAME"
.Vb 1
\&         tcpsocket - create a TCP socket and run a program
.Ve
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
tcpsocket [options] host port program [arguments]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtcpserver\fR creates a \s-1TCP\s0 socket, binds it to \fIhost\fR and \fIport\fR
and runs \fIprogram\fR with the remaining arguments.
.PP
\&\fIoptions\fR is a is a series of getopt or  getopt_long style options. See
below for more information.
.PP
The socket is described by \fIhost\fR and \fIport\fR. \fIhost\fR
may be 0, allowing connections to any local \s-1IP\s0 address; 
or a dotted-decimal \s-1IP\s0 address, allowing connections only to
that address; or a host name, allowing connections to the first \s-1IP\s0
address for that host. Host names are fed through qualification using
DNS_IP4_qualify.
\&\fIport\fR may be a name from /etc/services or a number; if it is 0,
\&\fBtcpsocket\fR will choose a free \s-1TCP\s0 port.
.SH "OPTIONS
Some people make this separate from the
description."
.IX Header "OPTIONS
Some people make this separate from the
description."
.Sh "Socket options:"
.IX Subsection "Socket options:"
.Ip "\-b, \-\-backlog=N" 8
.IX Item "-b, --backlog=N"
Allow a backlog of \fIN\fR \s-1TCP\s0 SYNs.
.Sp
On some systems, \fIN\fR is silently limited to 5.
On systems supporting \s-1SYN\s0 cookies, the backlog is
irrelevant. The default is 20.
.Ip "\-d, \-\-tcpdelay" 8
.IX Item "-d, --tcpdelay"
Delay sending data for a fraction of a second
whenever the remote host is responding slowly.
This is the default, but it may not be in the
future; if you want it, set it explicitly.
.Ip "\-D, \-\-tcpnodelay" 8
.IX Item "-D, --tcpnodelay"
Never delay sending data; enable \s-1TCP_NODELAY\s0.
.Ip "\-o, \-\-allow-ip-options" 8
.IX Item "-o, --allow-ip-options"
Leave \s-1IP\s0 options alone.
.Sp
If the client is sending packets along an
\&\s-1IP\s0 source route, send packets back along the
same route.
.Ip "\-O, \-\-kill-ip-options" 8
.IX Item "-O, --kill-ip-options"
Kill \s-1IP\s0 options.
.Sp
A client can still use source routing to connect
and to send data, but packets will be sent back
along the default route.
This is the default.
.Sh "Child program options:"
.IX Subsection "Child program options:"
.Ip "\-C, \-\-chroot=DIR" 8
.IX Item "-C, --chroot=DIR"
Change the working directory to \fI\s-1DIR\s0\fR. Then change the
root directory of the process to \fI.\fR (the current 
directory). 
This will be done after the socket has been created
and before changing the group or user \s-1ID\s0.
.Ip "\-F, \-\-fd=N" 8
.IX Item "-F, --fd=N"
Move socket handle to file descriptor \fIN\fR.
.Sp
\&\fIN\fR is an integer in the range from 0 to a system
"dependent limit. The default is to not move it.
Note that moving the socket to file descriptor 2
may disturb the output of error messages.
.Ip "\-g, \-\-gid=GID" 8
.IX Item "-g, --gid=GID"
Group \s-1ID\s0 to change to.
.Sp
tcpsocket will switch to that group id after
the socket has been set up. \s-1GID\s0 has to be a
positive integer.
.Ip "\-u, \-\-uid=UID" 8
.IX Item "-u, --uid=UID"
User \s-1ID\s0 to change to.
.Sp
tcpsocket will switch to that user id after
the socket has been set up. \s-1UID\s0 has to be a
positive integer.
.Ip "\-U, \-\-uidgid" 8
.IX Item "-U, --uidgid"
The same as \fB\-u\fR \f(CW$UID\fR \fB\-g\fR \f(CW$GID\fR.
.Sp
Usually used together with envuidgid.
.Sh "Data-gathering options:"
.IX Subsection "Data-gathering options:"
.Ip "\-l, \-\-localhost=NAME" 8
.IX Item "-l, --localhost=NAME"
Use \fI\s-1NAME\s0\fR as local host name.
.Sp
Do not look up the local host name in the \s-1DNS\s0.
This information is used for log files and
the environment variable \fIIODP_n_LOCALNAME\fR
.Ip "\-y, \-\-symbolic-name=NAME" 8
.IX Item "-y, --symbolic-name=NAME"
Use \fI\s-1NAME\s0\fR as symbolic socket name.
.Sp
This is used for \f(CW$IODP_n_SYMNAME\fR, and may be
used by \s-1IODP\s0 compliant application in an
application dependant way. The default is to not
set it.
.Sh "Verbosity options:"
.IX Subsection "Verbosity options:"
.Ip "\-1" 8
.IX Item "-1"
Print the local port number to the standard output.
.Ip "\-q, \-\-quiet" 8
.IX Item "-q, --quiet"
Quiet. Do not print error messages.
.Ip "\-Q" 8
.IX Item "-Q"
Print error messages.
.Sp
This is the default.
.Ip "\-v, \-\-verbose" 8
.IX Item "-v, --verbose"
Verbose. Print error and status messages.
.Ip "\*(--version" 8
.IX Item "version"
Show version information.
.Ip "\*(--help" 8
.IX Item "help"
Show a list of options or the long help on one.
.Sp
The use with an argument shows the long helptext
of that option, without an argument it will list
all options.
.Ip "\*(--longhelp" 8
.IX Item "longhelp"
Show longer help texts for all or one variable
.Sp
The use with an argument shows the long help for
this option, without arguments it shows the long
description of all options.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
\&\fBtcpsocket\fR sets the environment variables documented in the
\&\fIiodp-tcp\fR specification.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 5
\&        tcpsocket 0 119 \e
\&        setuidgid news \e
\&        env - PATH="$PATH:/news/bin" TMPDIR=/news/tmp SHELL=/bin/sh \e
\&            LOGNAME=news USER=news HOME=/news  \e
\&        /news/bin/innd
.Ve
.SH "BUGS"
.IX Header "BUGS"
unknown.
.SH "AUTHOR"
.IX Header "AUTHOR"
Uwe Ohse
.SH "SEE ALSO"
.IX Header "SEE ALSO"
udpsocket(8), uxstreamsocket(8), uxdgramsocket(8), file-iodp(8),
iodp(5).



NEWS
Version 0.2.4, January 2001

* added sample echo server.


tcpsocket.c
/*
 * Copyright (C) 2000 Uwe Ohse
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
#include "_stralloc.h"
#include "error.h"
#include "uogetopt.h"
#include "_alloc.h"
#include "fd.h"
#include "str2num.h"
#include "env.h"
#include "byte.h"
#include "fmt.h"
#include "bailout.h"
#include "ip4.h"
#include "socket.h"
#include "pathexec.h"
#include "iodptools.h"
#include "coe.h"

#define PROGNAME "tcpsocket"
const char *progname=PROGNAME;
int skipargs=3;

static const char *localhost;
static int opt_tcpnodelay=0;
static int opt_ipoptkill=1;

static unsigned long opt_backlog=20;

uogetopt_t localopts_socket[] = {
{0,"",UOGO_TEXT, 0,0, "Socket options:\n",0},
{'b', "backlog", UOGO_ULONG, &opt_backlog, 0,
"Allow a backlog of N TCP SYNs.\n"
"On some systems, n is silently limited to 5.\n"
"On systems supporting SYN cookies, the backlog is\n"
"irrelevant. The default is 20.","N"},
{'d',"tcpdelay", UOGO_FLAG, &opt_tcpnodelay, 0,
"Delay sending data for a fraction of a second\n"
"whenever the remote host is responding slowly.\n"
"This is the default, but it may not be in the\n"
"future; if you want it, set it explicitly.",0},
{'D',"tcpnodelay", UOGO_FLAG, &opt_tcpnodelay, 1,
"Never delay sending data; enable TCP_NODELAY.",0},
{'o',"allow-ip-options",UOGO_FLAG,&opt_ipoptkill,0,
"Leave IP options alone.\n"
"If the client is sending packets along an\n"
"IP source route, send packets back along the\n"
"same route.",0},
{'O',"kill-ip-options",UOGO_FLAG,&opt_ipoptkill,1,
"Kill IP options.\n"
"A client can still use source routing to connect\n"
"and to send data, but packets will be sent back\n"
"along the default route.\n"
"This is the default",0},
{0, 0, 0, 0, 0, 0, 0}
};
uogetopt_t localopts_data[] = {
{'l', "localhost", UOGO_STRING, &localhost, 0,
"Use NAME as local host name.\n"
"Do not look up the local host name in the DNS.\n"
"This information is used for log information and\n"
"$IODP_n_LOCALNAME.","NAME"},
{0, 0, 0, 0, 0, 0, 0}
};

const char *usagestr=
"usage: " PROGNAME " [options] host port program [arguments]";

int
doopen(char **argv)
{
	int sock;
	sock=prepare_socket4(argv[1],argv[2],1,1,opt_backlog,opt_print_port);
	if (opt_ipoptkill)
		if (-1==socket_ipoptionskill(sock)) {
			if (verbosity) warning(errno,"cannot disable IP options",0,0,0);
			_exit(111);
		}
	if (opt_tcpnodelay)
		if (-1==socket_tcpnodelay(sock)) {
			if (verbosity) warning(errno,"cannot enable TCPNODELAY",0,0,0);
			_exit(111);
		}
	return sock;
}

void
doiodp(int fd)
{
	iodp_set4(fd,localhost,"TCP4",opt_symname,"a",opt_backlog);
}
